examples/304-log-errors.js

```diff
@@ -0,0 +1,178 @@
+//
+// ===== AFTER =====
+// ERRO[0001] frame navigation to "https://test.k6.io/my_messages.php": timed out after 100ms
+//         at reflect.methodValueCall (native)
+//         at pageGotoTimeout100 (file:///home/ivan/Projects/grafana/xk6-browser/examples/304-log-errors.js:85:50(32))
+//         at native  executor=shared-iterations scenario=err source=stacktrace
+  page.goto('https://test.k6.io/my_messages.php', { timeout: 100 });
+
+  page.close();
+  browser.close();
+}
+
+export function pageQuery() {
+  const browser = launcher.launch('chromium', {
+    headless: __ENV.XK6_HEADLESS ? true : false,
+  });
+  const context = browser.newContext();
+  const page = context.newPage();
+
+  page.goto('https://test.k6.io/my_messages.php', { waitUntil: 'networkidle' });
+
+  // Enter login credentials and login
+  page.$('input[name="login"]').type('admin');
+  page.$('input[name="password"]').type('123');
+  page.$('input[type="submit"]').click();
+
+// RARE (race condition)
+// ===== BEFORE =====
+// ERRO[0003] error waiting for selector: element handle cannot evaluate: cannot call function on expression ("\n\t\t(node, injected, selector, strict, state, timeout, ...args) => {\n\t\t\treturn injected.waitForSelector(selector, node, strict, state, 'raf', timeout, ...args);\n\t\t}\n\t\n//# sourceURL=__xk6_browser_evaluation_script__\n") in execution context (3) in frame (BC46B592CD84782A8E5C9C0D8AFADA1A) with session (2178FD2416327DB14E218B78D7D051C9): Cannot find context with specified id (-32000)
+//         at reflect.methodValueCall (native)
+//         at pageWaitForSelector (file:///home/ivan/Projects/grafana/xk6-browser/examples/304-log-errors.js:132:29(70))
+//         at native  executor=shared-iterations scenario=err source=stacktrace
+//
+// ... or ...
+//
+// ERRO[0004] error waiting for selector: element handle cannot evaluate: cannot call function on expression ("\n\t\t(node, injected, selector, strict, state, timeout, ...args) => {\n\t\t\treturn injected.waitForSelector(selector, node, strict, state, 'raf', timeout, ...args);\n\t\t}\n\t\n//# sourceURL=__xk6_browser_evaluation_script__\n") in execution context (5) in frame (4CBD463ACA57362037F4E34B6367847C) with session (C034230EDD59875FB07D39F8A150AD71): exception "Uncaught (in promise)" (0:0): timed out after 0ms
+//         at reflect.methodValueCall (native)
+//         at pageWaitForSelector (file:///home/ivan/Projects/grafana/xk6-browser/examples/304-log-errors.js:132:29(70))
+//         at native  executor=shared-iterations scenario=err source=stacktrace
+//
+// ===== AFTER =====
+// ERRO[0008] querying selector "h5": execution context with ID 3 not found
+//         at reflect.methodValueCall (native)
+//         at pageQuery (file:///home/ivan/Projects/grafana/xk6-browser/examples/304-log-errors.js:125:15(70))
+//         at native  executor=shared-iterations scenario=err source=stacktrace
+  page.$('h5', { timeout: 0 });
```

Comment:
The new error message doesn't mention that `h5` couldn't be found, and instead mentions a mysterious `execution context with ID 3`. Would it be better to just say something like `timeout while looking for element h5: element not found`?

Reply:
The `execution context with ID 3 not found` is a race condition that we should fix, so it's not the same scenario as a timeout. It's a slight improvement over the previous `Cannot find context with specified id (-32000)` message, and since there's no friendlier way of logging this, we have to expose the internal [CDP concept of "execution contexts"](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#type-ExecutionContextDescription).

That said, I was wrong to mention that this can actually fail with a timeout. Notice that our [`Page.Query` method](https://github.com/grafana/xk6-browser/blob/f6b0363b12fdca200bb943a92a75c0dfacf91067/common/page.go#L673) doesn't actually receive an options object, and the [Playwright docs](https://playwright.dev/docs/api/class-page#page-query-selector) don't mention it either. So timeouts just aren't implemented for this method.

Reply:
If our goal is to make error messages user-friendly, even if we don't say "timeout", I [agree](https://github.com/grafana/xk6-browser/pull/429#issuecomment-1173665476) with @ankur22 that we might print a user-friendly message such as "not found" (or something in that sense) instead of the cryptic "execution context".

Reply:
But it's not cryptic :smile: Cryptic would be hiding the actual error to avoid confusing the user. This is something we need to fix so that the issue never happens, but saying anything else than "execution context not found" would be even more confusing.

If you have a specific suggestion let me know, but we can't just say "not found".

Reply:
The execution context error means that we couldn't find the element. Of course, we shouldn't hide the actual error; but I believe we should *translate it into the user's language*. The user shouldn't care whether we have an underlying race condition. "execution context with ID x not found" is confusing to users 😒

Like I explained [here](https://github.com/grafana/xk6-browser/pull/429#issuecomment-1173665476), we can tell the user we could not find the element and link to a document (in our repository) that explains the possible causes of this error _until we fix it_. By the way, there is no guarantee that it will never happen.

My two cents.

Reply:
I can't think of another cause other than a race condition 🤔 What we've seen so far, is this error only occurred because of an execution context switch, and we look for the element in a stale frame. Either the script code races with our internals, or we race with ourselves/CDP.

We can tell users in a linked issue/wiki that it might happen because of navigation and suggest possible solutions. If the proposed solutions won't fix their problem, we can advise them to open an issue or use the community forum.